From 373b297b131237850514fa5e7b46e1469ba38028 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:09:05 -0400
Subject: [usb] cdc-acm: fix open and suspend race

Message-id: <1410887463-20674-474-git-send-email-dzickus@redhat.com>
Patchwork-id: 95594
O-Subject: [RHEL7 PATCH 473/591] USB: cdc-acm: fix open and suspend race
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit 703df3297fb1950b0aa53e656108eb936d3f21d9
Author: Johan Hovold <jhovold@gmail.com>
Date:   Mon May 26 19:23:42 2014 +0200

    USB: cdc-acm: fix open and suspend race

    We must not do the usb_autopm_put_interface() before submitting the read
    urbs or we might end up doing I/O to a suspended device.

    Fixes: 088c64f81284 ("USB: cdc-acm: re-write read processing")
    Signed-off-by: Johan Hovold <jhovold@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/class/cdc-acm.c b/drivers/usb/class/cdc-acm.c
index c255e77..74df311 100644
--- a/drivers/usb/class/cdc-acm.c
+++ b/drivers/usb/class/cdc-acm.c
@@ -528,19 +528,15 @@ static int acm_port_activate(struct tty_port *port, struct tty_struct *tty)
  if (usb_submit_urb(acm->ctrlurb, GFP_KERNEL)) {
   dev_err(&acm->control->dev,
    "%s - usb_submit_urb(ctrl irq) failed\n", __func__);
-  usb_autopm_put_interface(acm->control);
   goto error_submit_urb;
  }
 
  acm->ctrlout = ACM_CTRL_DTR | ACM_CTRL_RTS;
  if (acm_set_control(acm, acm->ctrlout) < 0 &&
      (acm->ctrl_caps & USB_CDC_CAP_LINE)) {
-  usb_autopm_put_interface(acm->control);
   goto error_set_control;
  }
 
- usb_autopm_put_interface(acm->control);
-
  /*
   * Unthrottle device in case the TTY was closed while throttled.
   */
@@ -552,6 +548,8 @@ static int acm_port_activate(struct tty_port *port, struct tty_struct *tty)
  if (acm_submit_read_urbs(acm, GFP_KERNEL))
   goto error_submit_read_urbs;
 
+ usb_autopm_put_interface(acm->control);
+
  mutex_unlock(&acm->mutex);
 
  return 0;
@@ -562,6 +560,7 @@ error_submit_read_urbs:
 error_set_control:
  usb_kill_urb(acm->ctrlurb);
 error_submit_urb:
+ usb_autopm_put_interface(acm->control);
 error_get_interface:
 disconnected:
  mutex_unlock(&acm->mutex);
-- 
1.7.1